rpc: improve getaddressinfo test coverage, help, code docs

jonatack opened the pull request:
This PR is a continuation of the work in https://github.com/bitcoin/bitcoin/pull/12892.

Main motivations:
- There is currently no test coverage for the getaddressinfo `labels` response. Coverage here is a prerequisite before deprecating the `label` response or adding multiple labels per address.
- `bitcoin-cli help getaddressinfo` returns a few content errors, difficult-to-read formatting, and no explanation why it returns both `label` and `labels` and how they relate, which can be confusing for application developers.

Changes by order of commits:
- [x] improve/fix getaddressinfo RPCHelpman layout formatting
- [x] improve/fix getaddressinfo RPCHelpman content
- [x] clarify the `label` and `labels` fields in getaddressinfo RPCHelpman
- [x] update getaddressinfo RPCExamples addresses to bech32
- [x] add getaddressinfo code docs
- [x] add a `listlabels` test assertion in wallet_labels.py
- [x] add missing getaddressinfo `labels` test coverage and improve the existing `label` tests

Here are gists of the CLI help output:
[`bitcoin-cli help getaddressinfo` before this PR](https://gist.github.com/jonatack/022af5221a85c069780359a22643c810)
[`bitcoin-cli help getaddressinfo` after this PR](https://gist.github.com/jonatack/4ee5f6abc62a3d99269570206a5f90ba)

It seems we ought to begin a deprecation process for the getaddressinfo `label` field? If yes, I have a follow-up ready. _--> EDIT: Deprecation follow-ups #17578 and #17585 now build on this PR._

jonatack committed:
rpc: improve getaddressinfo RPCHelpman formatting

jonatack committed:
rpc: improve getaddressinfo RPCHelpman content

jonatack committed:
rpc: add getaddressinfo code documentation

and separate the fields with a line break for readability.

jonatack committed:
test: add rpc getaddressinfo labels test coverage

jonatack committed:
test: add listlabels test in wallet_labels.py

jonatack committed:
rpc: clarify label vs labels in getaddressinfo RPCHelpman

jonatack committed:
rpc: update getaddressinfo RPCExamples to bech32

jachiang commented on test/functional/test_framework/wallet_util.py:
```diff
@@ -88,6 +88,11 @@ def get_multisig(node):
                     p2sh_p2wsh_script=CScript([OP_HASH160, witness_script, OP_EQUAL]).hex(),
                     p2sh_p2wsh_addr=script_to_p2sh_p2wsh(script_code))
 
+def labels_value(name="", purpose="receive"):
```
Question: Would it make sense for the `labels_value` method to indicate that it is only for the single member `labels` case (since `getaddressinfo` may support multiple labels in the future)?

jonatack replied:
Thanks @jachiang. I thought I'd update it if/when adding support for multiple labels, though happy to take a good suggestion now for a better name or docstring.

jachiang replied:
You're right, a docstring or inline comment would have probably sufficient for sb who is just looking at the functional test, and not the rpc code (where labels vector is actually documented). Happy to wait and see what other reviewers think :)

jonatack replied:
The rpc code doc seems good, but explaining it in the help might be a good idea. Thanks, might update that. Thanks for the review!

jachiang reviewed (approved):
tACK d73709a9decdb34562de9c847eee786e6a256a82 
Thanks @jonatack. 
I reviewed diffs, ran passing tests.

fjahr commented on test/functional/wallet_basic.py:
```diff
@@ -47,9 +50,6 @@ def check_fee_amount(self, curr_balance, balance_with_fee, fee_per_byte, tx_size
     def get_vsize(self, txn):
         return self.nodes[0].decoderawtransaction(txn)['vsize']
 
-    def labels_value(self, name="", purpose="receive"):
```
Would combine this commit with bb87d8aac71dc87d1f11c26bd97fa6dc7b650a75, as you are mostly touching LoC that you touched in the that one as well.

fjahr commented on test/functional/wallet_importmulti.py:
```diff
@@ -119,10 +119,12 @@ def run_test(self):
         # ScriptPubKey + internal + label
         self.log.info("Should not allow a label to be specified when internal is true")
         key = get_key(self.nodes[0])
+        label = "Unsuccessful labelling for internal addresses"
```
I think this one is touching the same lines as 3dcf1481cefae6ff2d585181578004045c3e3a1c, so I would combine these commits as well.

jonatack replied:
@fjahr I squashed the 4 commits adding RPC getaddressinfo "labels" test coverage and improving the existing "label" tests down to a single commit at e123923e630c051f676501aab9454ecbf3bd3956. Can you please re-ACK?

fjahr reviewed (approved):
Code review ACK d73709a

Good cleanup. I would just recommend compressing the commits as indicated in my comments. I felt like having to review the same lines multiple times as I went through the commits and combining them will cut down on time needed for reviewing.

jonatack reviewed (commented):
Thank you @jachiang and @fjahr for your reviews. Rebased and added more detail to the label and labels help in 235417d and code docs in 6b20250.

fjahr commented:
Code review ACK e123923e630c051f676501aab9454ecbf3bd3956

ariard reviewed (approved):
Code review and tested ACK e123923 

IMO not sure it's worth it splitting getaddressinfo content and formatting change in diff commits.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3709,15 +3709,17 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
             "  \"iscompressed\" : true|false,        (boolean, optional) If the pubkey is compressed.\n"
-            "  \"label\" :  \"label\"                  (string) The label associated with the address. Defaults to \"\".\n"
+            "  \"label\" :  \"label\"                  (string) The label associated with the address. Defaults to \"\". Equivalent to the name field in the labels array\n"
+            "                                               below. This label field may be deprecated in the future; prefer consuming the labels array below instead.\n"
```
Yes. This is merely a recommandation in the help as a helpful guide for application developers confused whether to use `label` and `labels`. Happy to remove it if the roadmap is to not deprecate `label` at some point.

jonatack replied:
Thanks for the review @ariard!

luke-jr replied:
It seems unclear if multiple labels makes sense, so I don't think we're ready to deprecate this yet.

(Multiple tags does make sense, but "label" has historically been a description for the specific tx)

jonatack replied:
Removed the line about deprecation.
```diff
-"\"label\" :  \"label\"   (string) The label associated with the address. Defaults to \"\". Equivalent to the name field in the labels array\n"
-"                              below. This label field may be deprecated in the future; prefer consuming the labels array below instead.\n"
+"\"label\" :  \"label\"   (string) The label associated with the address. Defaults to \"\". Equivalent to the name field in the labels array.\n"
```

jachiang commented:
Code review ACK e123923e630c051f676501aab9454ecbf3bd3956

I find the clarification on `Label` vs. `Labels` in comments/help and the newly added test coverage of `Labels` helpful for anybody interfacing with `getaddressinfo`. Use `Labels` :)

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3736,32 +3736,56 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
     CTxDestination dest = DecodeDestination(request.params[0].get_str());
 
-    // Make sure the destination is valid
+    // Ensure the destination is valid.
```
1d6884d4d4b541d0c04ef10da208ef657122aa1f

The code below is clear and simple, why add the obvious comment?

jonatack replied:
Removed.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3736,32 +3736,56 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid address");
     }
 
+    // Return address field.
```
1d6884d4d4b541d0c04ef10da208ef657122aa1f

IMO these "return foo field" comments are useless.

jonatack replied:
Thank you for reviewing, @promag. Some of the comments add info that isn't obvious and the others are for consistency and separation. I propose to keep the 3 tested ACKs here and remove too-obvious comments in the follow-up (to add multiple labels / deprecate the label field).

jonatack replied:
Had to rebase, so removed the obvious comments.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3785,16 +3810,19 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
     }
     ret.pushKV("labels", std::move(labels));
 
+    // Return final response.
```
1d6884d4d4b541d0c04ef10da208ef657122aa1f

Pretty obvious too?

jonatack replied:
Removed

promag reviewed (commented):
Overall looks good, only read the change so far.

jonatack commented:
Thank you @jachiang and @promag for reviewing.

michaelfolkson commented:
tACK e123923e630c051f676501aab9454ecbf3bd3956. CLI help output as described. Looking forward to reviewing the deprecation of label PR ;)

jonatack commented:
Rebased.

jonatack commented:
appveyor failures are unrelated:
```
bech32_tests.obj : error LNK2001: unresolved external symbol "bool __cdecl CaseInsensitiveEqual
```

ryanofsky commented:
> appveyor failures are unrelated:

Appveyor error is apparently fixed by https://github.com/bitcoin/bitcoin/pull/17384#issuecomment-550305021

jonatack commented:
Rebased and incorporated @promag's feedback.

jnewbery commented:
Concept ACK. Thanks Jon.

I find the `purpose` field in the `labels` return object very strange, and think that we should probably remove it before encouraging people to rely on `labels`. The code to add the purpose field:

https://github.com/bitcoin/bitcoin/blob/21ee676dd6a7d9704367b6412bf8e1e443ec2b5b/src/wallet/rpcwallet.cpp#L3672

was added in #12892 (from code originally in #7729). Prior to that, the `purpose` field of the `CAddressBookData` object wasn't exposed in the rpc.

My preference would be to:

- make the `label` return object deprecated (only return it if bitcoind is started with a specific `deprecatedrpc` flag).
- also make the `purpose` field of the labels object deprecated (only return it if bitcoind is start with the same `deprecatedrpc` flag) and instead return a flat array of label strings.
- fully remove the `label` return value and `purpose` field in v0.21.

Some feedback on this PR:

- I personally wouldn't reference twitter or mastodon in commit messages (since I expect the bitcoind git repo to last longer than those URLs). Also the justification in that commit log seems like an appeal to authority (this experienced developer found it confusing) rather than justifying the change on its own merit
- in the last commit, no need to say that this is squashed. Once this is merged, then the git history will only include that commit. There's no need to say how that commit came into being.
- no need to update dates in the copyright notices. That just increases the diff unnecessarily. We update them all automatically at the end of the year.
- if you do agree with me and think that `purpose` should be removed, then there's not much point in adding the the `labels_value()` function in wallet_util.py.

jonatack commented:
Thank you for the review, @jnewbery! I took your suggestions: removed mention of upcoming deprecation of the label field from the rpc help, updated the PR description and commit messages, and removed the copyright header updates. 

I think this PR now does what is needed for the current situation while making minimum changes WRT existing review and is hopefully RFM.

>  if you do agree with me and think that `purpose` should be removed, then there's not much point in adding the the `labels_value()` function in wallet_util.py.

Since I don't know if it will be accepted, I'd like to propose the `labels purpose` deprecation in another PR. Idem for deprecating the `label` field and for adding multiple labels (tags) capability.

jonatack commented:
Found and fixed another content error in the help text, and rebased the change into af63c0a _rpc: improve getaddressinfo RPCHelpman content_
```diff
-"  \"pubkeys\"         (string, optional) Array of pubkeys associated with the known redeemscript (only if script is multisig).\n"
+"  \"pubkeys\"         (array, optional) Array of pubkeys associated with the known redeemscript (only if script is multisig).\n"
             "    [\n"
-"      \"pubkey\"\n"
+"      \"pubkey\" (string)\n"
```

jonatack commented:
Deprecated getaddressinfo labels `purpose` and JSON hash object in PR #17578.

jonatack commented:
Deprecated the getaddressinfo `label` field in #17585 which builds on this PR.

fjahr commented:
Re-ACK 33f5fc32e5bfbe1e89c4d20ce455bcc6dc194151

Reviewed code changes. Build failure seems unrelated, would need a rebuild.

jonatack commented:
4 ACKs and 1-2 concept ACKs for functional tests and docs. Is anything else needed here?

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3773,23 +3778,39 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
     ret.pushKV("iswatchonly", bool(mine & ISMINE_WATCH_ONLY));
+
+    // Return DescribeWalletAddress fields.
+    // Always returned: isscript, ischange, iswitness.
```
nit (please don't invalidate review ACKs to change this): this commenting would be better placed next to the `DescribeWalletAddress` function or `DescribeWalletAddressVisitor` class, since it's very possible that the key-values returned will change in future, and keeping comments close to the code it's describing reduce the chances of them going stale.

jonatack replied:
Good idea -- will update #17578 with this.

jonatack replied:
Removed.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3773,23 +3778,39 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+
     ret.pushKV("ischange", pwallet->IsChange(scriptPubKey));
 
+    // Fetch KeyMetadata, if present, for the timestamp, hdkeypath, hdseedid,
```
nit: Seems redundant. The code below is very easy to read.

jonatack replied:
ACK, will update #17578 with this.

jonatack replied:
Removed in #17578.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3709,52 +3709,57 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
                 RPCExamples{
-                    HelpExampleCli("getaddressinfo", "\"1PSSGeFHDnKNxiEyFrD1wcEaHr9hrQDDWc\"")
-            + HelpExampleRpc("getaddressinfo", "\"1PSSGeFHDnKNxiEyFrD1wcEaHr9hrQDDWc\"")
+                    HelpExampleCli("getaddressinfo", "\"bc1q09vm5lfy0j5reeulh4x5752q25uqqvz34hufdl\"") +
```
Is this intentionally an invalid address? The previous example was a valid address, which I think is useful since users can test the RPC and see a valid return object by using the example

jonatack replied:
Yes, I noticed addresses with a missing character in other RPC examples and thought it might be the norm. Will update #17578.

jonatack replied:
Added a valid address in #17578.

jnewbery reviewed (commented):
ACK 33f5fc32e5bfbe1e89c4d20ce455bcc6dc194151.

I've left some nits. please don't invalidate review now, but you could consider addressing them in your follow-up PR.

sidhujag commented:
This seemed to break CI / Trusty

jonatack commented:
> This seemed to break CI / Trusty or at the very least, Trusty distribution seemed to not be able to be fetched from here on.

@sidhujag I think it's unrelated; see #17626.
